presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushProjectionIntoTableScan.java

```diff
@@ -0,0 +1,130 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.TableHandle;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.ProjectionApplicationResult;
+import io.prestosql.spi.expression.ConnectorExpression;
+import io.prestosql.spi.expression.ConnectorExpressionTranslator;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.tree.Expression;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.tableScan;
+
+public class PushProjectionIntoTableScan
+        implements Rule<ProjectNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+    private static final Pattern<ProjectNode> PATTERN = project().with(source().matching(
+            tableScan().capturedAs(TABLE_SCAN)));
+
+    private final Metadata metadata;
+    private final TypeAnalyzer typeAnalyzer;
+
+    public PushProjectionIntoTableScan(Metadata metadata, TypeAnalyzer typeAnalyzer)
+    {
+        this.metadata = metadata;
+        this.typeAnalyzer = typeAnalyzer;
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(ProjectNode project, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        List<ConnectorExpression> projections;
+        try {
+            projections = project.getAssignments()
+                    .getExpressions().stream()
+                    .map(expression -> ConnectorExpressionTranslator.translate(
+                            context.getSession(),
+                            expression,
+                            typeAnalyzer,
+                            context.getSymbolAllocator().getTypes()
+                    ))
+                    .collect(Collectors.toList());
+        }
+        catch (UnsupportedOperationException e) {
+            // some expression not supported by translator, skip
```

Comment:
LOG.debug?

Reply:
TODO: add support for partial partial pushdown?

Reply:
LOG.debug may end up being too verbose in the short term. Think of this as a poor-man's "matcher". It only matches when an expression can be translated.